spyder/api/plugins.py

```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+    FONT_SIZE_DELTA = 0
+    RICH_FONT_SIZE_DELTA = 0
+
+    # --- API: Signals -------------------------------------------------------
+    # ------------------------------------------------------------------------
+
+    # Use this signal to inform another plugin that configuration changed.
+    sig_free_memory_requested = Signal()
+    sig_option_changed = Signal(str, object)
+    sig_quit_requested = Signal()
+    sig_redirect_stdio_requested = Signal(bool)
+    sig_restart_requested = Signal()
+    sig_shortcut_registration_requested = Signal(QObject, str, str, bool)
+    sig_status_message_requested = Signal(str, int)
+
+    # --- Private attributes -------------------------------------------------
+    # ------------------------------------------------------------------------
+    # Define configuration name map for plugin to split configuration
+    # among several files. See spyder/config/main.py
+    _CONF_NAME_MAP = None
+
+    # Store reference to registered plugins in a dictionary
+    # {plugin.NAME: plugin_instance, ...}
+    _PLUGINS = None
+
+    def __init__(self, parent, configuration=None):
+        super().__init__(parent)
+
+        self._main = parent
+        self.main = parent
+        self._conf = configuration
+        self._widget = None
+        self._plugin_path = os.path.dirname(inspect.getfile(self.__class__))
+        self._is_registered = None
+        self._is_compatible = None
+
+        self._actions = {}
+
+    # --- Private methods ---------------------------------------------
+    # ------------------------------------------------------------------------
+    def _register(self):
+        """
+        Setup and register plugin in Spyder's main window and connect it to
+        other plugins.
+        """
+        self.WIDGET_CLASS if self.WIDGET_CLASS else None
+
+        # Checks
+        # --------------------------------------------------------------------
+        if self._PLUGINS is None:
+            self._PLUGINS = {}
+
+        if self.NAME is None:
+            raise Exception('A Spyder Plugin must define a `NAME`!')
+
+        if self.NAME in self._PLUGINS:
+            raise Exception('A Spyder Plugin with NAME="{}" already exists!'
+                            ''.format(self.NAME))
+
+        # Setup configuration
+        # --------------------------------------------------------------------
+        self._PLUGINS[self.NAME] = self
+        if self._conf is not None:
+            self._conf.register_plugin(self)
+
+        # Signals
+        # --------------------------------------------------------------------
+        self._widget.sig_option_changed.connect(self.sig_option_changed)
+        self.sig_option_changed.connect(self.set_option)
+        self._is_registered = True
+
+    def _unregister(self):
+        """
+        Disconnect signals and clean up the plugin to be able to stop a plugin
+        while Spyder is running.
+        """
+        self.sig_option_changed.disconnect()
```

Comment:
This probably needs to be inside a try/except TypeError. This full `disconnect` frequently generates that error.

Reply:
Fixed